pennylane/templates/subroutines/amplitude_amplification.py

```diff
@@ -0,0 +1,184 @@
+
+    def _flatten(self):
+        data = (self.hyperparameters["U"], self.hyperparameters["O"])
+        metadata = tuple(
+            value for key, value in self.hyperparameters.items() if key not in ["O", "U"]
+        )
+        return data, metadata
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        U, O = (data[0], data[1])
+        return cls(
+            U,
+            O,
+            iters=metadata[0],
+            fixed_point=metadata[1],
+            work_wire=metadata[2],
+            p_min=metadata[3],
+            reflection_wires=metadata[4],
+        )
+
+    def __init__(
+        self, U, O, iters=1, fixed_point=False, work_wire=None, p_min=0.9, reflection_wires=None
+    ):
+        self._name = "AmplitudeAmplification"
+        if reflection_wires is None:
+            reflection_wires = U.wires
+
+        if fixed_point and work_wire is None:
+            raise qml.wires.WireError("work_wire must be specified if fixed_point == True.")
+
+        if fixed_point and len(O.wires + qml.wires.Wires(work_wire)) == len(O.wires):
+            raise ValueError("work_wire must be different from the wires of O.")
+
+        if fixed_point:
+            wires = U.wires + qml.wires.Wires(work_wire)
+        else:
+            wires = U.wires
+
+        self.hyperparameters["U"] = U
+        self.hyperparameters["O"] = O
+        self.hyperparameters["iters"] = iters
+        self.hyperparameters["fixed_point"] = fixed_point
+        self.hyperparameters["work_wire"] = work_wire
+        self.hyperparameters["p_min"] = p_min
+        self.hyperparameters["reflection_wires"] = qml.wires.Wires(reflection_wires)
+
+        super().__init__(wires=wires)
+
+    # pylint:disable=arguments-differ
+    @staticmethod
+    def compute_decomposition(**kwargs):
+        U = kwargs["U"]
+        O = kwargs["O"]
+        iters = kwargs["iters"]
+        fixed_point = kwargs["fixed_point"]
+        work_wire = kwargs["work_wire"]
+        p_min = kwargs["p_min"]
+        reflection_wires = kwargs["reflection_wires"]
+
+        ops = []
+
+        if fixed_point:
+            alphas, betas = _get_fixed_point_angles(iters, p_min)
+
+            for iter in range(iters // 2):
+                ops.append(qml.Hadamard(wires=work_wire))
+                ops.append(qml.ctrl(O, control=work_wire))
+                ops.append(qml.Hadamard(wires=work_wire))
+                ops.append(qml.PhaseShift(betas[iter], wires=work_wire))
+                ops.append(qml.Hadamard(wires=work_wire))
+                ops.append(qml.ctrl(O, control=work_wire))
+                ops.append(qml.Hadamard(wires=work_wire))
+
+                ops.append(qml.Reflection(U, -alphas[iter], reflection_wires=reflection_wires))
+        else:
+            for _ in range(iters):
+                ops.append(O)
+                ops.append(qml.Reflection(U, np.pi, reflection_wires=reflection_wires))
+
+        return ops
+
+    # pylint: disable=protected-access
```

Comment:
Is this still needed? We aren't accessing any protected variables

Reply:
removed